core-features/auto-recovery/index.test.js

```diff
@@ -0,0 +1,130 @@
+    })
+    try {
+      await client.connect()
+      await client.query(`
+      DROP TYPE IF EXISTS "Role";
+      CREATE TYPE "Role" AS ENUM ('USER', 'ADMIN');
+      DROP TABLE IF EXISTS "public"."Post" CASCADE;
+      CREATE TABLE "public"."Post" (
+          "id" text NOT NULL,
+          "createdAt" timestamp(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
+          "updatedAt" timestamp(3) NOT NULL DEFAULT '1970-01-01 00:00:00'::timestamp without time zone,
+          "published" boolean NOT NULL DEFAULT false,
+          "title" text NOT NULL,
+          "content" text,
+          "authorId" text,
+          "jsonData" jsonb,
+          PRIMARY KEY ("id")
+      );
+      DROP TABLE IF EXISTS "public"."User" CASCADE;
+      CREATE TABLE "public"."User" (
+          "id" text,
+          "email" text NOT NULL,
+          "name" text,
+          PRIMARY KEY ("id")
+      );
+      CREATE UNIQUE INDEX "User.email" ON "public"."User"("email");
+      ALTER TABLE "public"."Post" ADD FOREIGN KEY ("authorId") REFERENCES "public"."User"("id")
+      ON DELETE SET NULL ON UPDATE CASCADE;
+      INSERT INTO "public"."User" (email, id, name) VALUES ('a@a.de',	'576eddf9-2434-421f-9a86-58bede16fd95',	'Alice');
+    `)
+      prismaClient = new PrismaClient({
+        errorFormat: 'colorless',
+        __internal: {
+          measurePerformance: true,
+        },
+        datasources: {
+          db: {
+            url: proxyConnectionString.href,
+          },
+        },
+        log: [
+          {
+            emit: 'event',
+            level: 'error',
+          },
+        ],
+      })
+    } catch (e) {
+      console.log(e)
+    }
+  })
+
+  afterAll(async () => {
+    await prismaClient.$disconnect()
+    await client.query(`
+    DROP TABLE IF EXISTS "public"."Post" CASCADE;
+    DROP TABLE IF EXISTS "public"."User" CASCADE;`)
+    await client.end()
+  })
+
+  it('should verify that the prisma client automatically recovers after an interrupted db server connection', async () => {
+    const proxy = tcpProxy.createProxy(newPort, hostname, port, {})
+
+    // confirm connection is working generally
+    await prismaClient.$connect() 
+    try {
+      await prismaClient.user.findMany()
+    } catch (e) {
+      if (!(e instanceof PrismaClientValidationError)) { 
+        throw new Error(`Validation error is incorrect`)
+      }
+      errorLogs.push(e)
+    }
+    expect(errorLogs.length).toBe(0)
+    await prismaClient.$disconnect()
+    await prismaClient.$connect()
+    proxy.end()
+    try {
+      await prismaClient.user.findMany()
+    } catch (e) {
+      if (!(e instanceof PrismaClientKnownRequestError)) { // should throw P1001 - Can't reach database server 
+        throw new Error(`Error code is incorrect`)
+      }
+      errorLogs.push(e)
+    }
+    expect(errorLogs.length).toBe(1)
+    const proxy2 = tcpProxy.createProxy(newPort, hostname, port, {})
+    await new Promise((r) => setTimeout(r, 16000)) 
```

Comment:
I noticed that the timeout here, initially added in consideration of the 15s health check that the connection pooler is supposed to do automatically, is not necessary for the query immediately after the port is reopened to succeed. No new connect() or timeout required. 
Since the timeout is no longer needed, and the fact that we're really only testing prisma client recovery rather than the whole health check process, maybe this was actually fine to live inside the integration test folder after all?

Reply:
That is interesting - because that means this test is testing something else than what it _should_ test.

Reply:
I might have a hunch why this is working without the timeout - right now it possibly just opens another connection to the database. You might want to limit the amount of connections the Query Engine can open to 1 (connection limit = 1). That should force a reuse of the connection that is now broken and hopefully show the behavior we expect.

Reply:
Just tried this in my last commit [9cbc6ad] and it still succeeds without the timeout :/ (when I run the test locally)
I also moved the test postgres service into the auto-recovery folder for now (just seen that the CI test is failing though - missed something in run.sh apparently).

Reply:
(I wrote the original test)
I think Jan's theory is correct here, we need to limit the number of connections to 1 by adding `&connection_limit=1` to the connection string.